
This project aims to analyze and compare the effectiveness of two digital marketing campaigns

In [44]:
#IMPORTING LIBRARIES
import numpy as np
import pandas as pd
from scipy.stats import ttest_1samp,shapiro,levene,ttest_ind,mannwhitneyu, \
    pearsonr,spearmanr,kendalltau,f_oneway,kruskal
from scipy.stats import mannwhitneyu


In [45]:
#importing the data
df_control=pd.read_csv("control_group.csv",sep=";")
df_control.head()

Campaign Name       Date  Spend [USD]  # of Impressions     Reach  \
0  Control Campaign  1.08.2019         2280           82702.0   56930.0   
1  Control Campaign  2.08.2019         1757          121040.0  102513.0   
2  Control Campaign  3.08.2019         2343          131711.0  110862.0   
3  Control Campaign  4.08.2019         1940           72878.0   61235.0   
4  Control Campaign  5.08.2019         1835               NaN       NaN   

   # of Website Clicks  # of Searches  # of View Content  # of Add to Cart  \
0               7016.0         2290.0             2159.0            1819.0   
1               8110.0         2033.0             1841.0            1219.0   
2               6508.0         1737.0             1549.0            1134.0   
3               3065.0         1042.0              982.0            1183.0   
4                  NaN            NaN                NaN               NaN   

   # of Purchase  
0          618.0  
1          511.0  
2          372.0  
3          340.0  
4            NaN

In [46]:
df_test=pd.read_csv("test_group.csv",sep=";")
df_test.head()

Campaign Name       Date  Spend [USD]  # of Impressions  Reach  \
0  Test Campaign  1.08.2019         3008             39550  35820   
1  Test Campaign  2.08.2019         2542            100719  91236   
2  Test Campaign  3.08.2019         2365             70263  45198   
3  Test Campaign  4.08.2019         2710             78451  25937   
4  Test Campaign  5.08.2019         2297            114295  95138   

   # of Website Clicks  # of Searches  # of View Content  # of Add to Cart  \
0                 3038           1946               1069               894   
1                 4657           2359               1548               879   
2                 7885           2572               2367              1268   
3                 4216           2216               1437               566   
4                 5863           2106                858               956   

   # of Purchase  
0            255  
1            677  
2            578  
3            340  
4            768

In [47]:
#renaming columns for better understanding and calling
df_control.columns = ["Campaign Name", "Date", "Amount Spent", 
                        "Number of Impressions", "Reach", "Website Clicks", 
                        "Searches Received", "Content Viewed", "Added to Cart",
                        "Purchases"]

df_test.columns = ["Campaign Name", "Date", "Amount Spent", 
                        "Number of Impressions", "Reach", "Website Clicks", 
                        "Searches Received", "Content Viewed", "Added to Cart",
                        "Purchases"]

In [48]:
#sizes of each campaign's sample
df_control.shape, df_test.shape

((30, 10), (30, 10))

In [49]:
df_control.describe()

Amount Spent  Number of Impressions          Reach  Website Clicks  \
count     30.000000              29.000000      29.000000       29.000000   
mean    2288.433333          109559.758621   88844.931034     5320.793103   
std      367.334451           21688.922908   21832.349595     1757.369003   
min     1757.000000           71274.000000   42859.000000     2277.000000   
25%     1945.500000           92029.000000   74192.000000     4085.000000   
50%     2299.500000          113430.000000   91579.000000     5224.000000   
75%     2532.000000          121332.000000  102479.000000     6628.000000   
max     3083.000000          145248.000000  127852.000000     8137.000000   

       Searches Received  Content Viewed  Added to Cart   Purchases  
count          29.000000       29.000000      29.000000   29.000000  
mean         2221.310345     1943.793103    1300.000000  522.793103  
std           866.089368      777.545469     407.457973  185.028642  
min          1001.000000      848.000000     442.000000  222.000000  
25%          1615.000000     1249.000000     930.000000  372.000000  
50%          2390.000000     1984.000000    1339.000000  501.000000  
75%          2711.000000     2421.000000    1641.000000  670.000000  
max          4891.000000     4219.000000    1913.000000  800.000000

In [50]:
df_test.describe()

Amount Spent  Number of Impressions          Reach  Website Clicks  \
count     30.000000              30.000000      30.000000       30.000000   
mean    2563.066667           74584.800000   53491.566667     6032.333333   
std      348.687681           32121.377422   28795.775752     1708.567263   
min     1968.000000           22521.000000   10598.000000     3038.000000   
25%     2324.500000           47541.250000   31516.250000     4407.000000   
50%     2584.000000           68853.500000   44219.500000     6242.500000   
75%     2836.250000           99500.000000   78778.750000     7604.750000   
max     3112.000000          133771.000000  109834.000000     8264.000000   

       Searches Received  Content Viewed  Added to Cart   Purchases  
count          30.000000       30.000000      30.000000   30.000000  
mean         2418.966667     1858.000000     881.533333  521.233333  
std           388.742312      597.654669     347.584248  211.047745  
min          1854.000000      858.000000     278.000000  238.000000  
25%          2043.000000     1320.000000     582.500000  298.000000  
50%          2395.500000     1881.000000     974.000000  500.000000  
75%          2801.250000     2412.000000    1148.500000  701.000000  
max          2978.000000     2801.000000    1391.000000  890.000000

# Missing values

In [51]:
df_test.isnull().sum()

Campaign Name            0
Date                     0
Amount Spent             0
Number of Impressions    0
Reach                    0
Website Clicks           0
Searches Received        0
Content Viewed           0
Added to Cart            0
Purchases                0
dtype: int64

In [52]:
df_control.isnull().sum()

Campaign Name            0
Date                     0
Amount Spent             0
Number of Impressions    1
Reach                    1
Website Clicks           1
Searches Received        1
Content Viewed           1
Added to Cart            1
Purchases                1
dtype: int64

There is one entry missing in control group in a few columns.                                                          

In [53]:
#filling the empty entries with the mean of the column:
df_control.fillna({"Number of Impressions": df_control["Number of Impressions"].mean()}, inplace=True)
df_control.fillna({"Reach": df_control["Reach"].mean()}, inplace=True)
df_control.fillna({"Website Clicks": df_control["Website Clicks"].mean()}, inplace=True)
df_control.fillna({"Searches Received": df_control["Searches Received"].mean()}, inplace=True)
df_control.fillna({"Added to Cart": df_control["Added to Cart"].mean()}, inplace=True)
df_control.fillna({"Purchases": df_control["Purchases"].mean()}, inplace=True)
df_control.fillna({"Content Viewed": df_control["Content Viewed"].mean()}, inplace=True)

df_control.isnull().sum()

Campaign Name            0
Date                     0
Amount Spent             0
Number of Impressions    0
Reach                    0
Website Clicks           0
Searches Received        0
Content Viewed           0
Added to Cart            0
Purchases                0
dtype: int64

In [54]:
#Merging
df=pd.concat([df_control, df_test], ignore_index=True)
print(df.head())
# df.head()
df.tail()


      Campaign Name       Date  Amount Spent  Number of Impressions  \
0  Control Campaign  1.08.2019          2280           82702.000000   
1  Control Campaign  2.08.2019          1757          121040.000000   
2  Control Campaign  3.08.2019          2343          131711.000000   
3  Control Campaign  4.08.2019          1940           72878.000000   
4  Control Campaign  5.08.2019          1835          109559.758621   

           Reach  Website Clicks  Searches Received  Content Viewed  \
0   56930.000000     7016.000000        2290.000000     2159.000000   
1  102513.000000     8110.000000        2033.000000     1841.000000   
2  110862.000000     6508.000000        1737.000000     1549.000000   
3   61235.000000     3065.000000        1042.000000      982.000000   
4   88844.931034     5320.793103        2221.310345     1943.793103   

   Added to Cart   Purchases  
0         1819.0  618.000000  
1         1219.0  511.000000  
2         1134.0  372.000000  
3         1183.0  340.

Campaign Name        Date  Amount Spent  Number of Impressions    Reach  \
55  Test Campaign  26.08.2019          2311                80841.0  61589.0   
56  Test Campaign  27.08.2019          2915               111469.0  92159.0   
57  Test Campaign  28.08.2019          2247                54627.0  41267.0   
58  Test Campaign  29.08.2019          2805                67444.0  43219.0   
59  Test Campaign  30.08.2019          1977               120203.0  89380.0   

    Website Clicks  Searches Received  Content Viewed  Added to Cart  \
55          3820.0             2037.0          1046.0          346.0   
56          6435.0             2976.0          2552.0          992.0   
57          8144.0             2432.0          1281.0         1009.0   
58          7651.0             1920.0          1240.0         1168.0   
59          4399.0             2978.0          1625.0         1034.0   

    Purchases  
55      284.0  
56      771.0  
57      721.0  
58      677.0  
59      572.0

In [55]:
df.groupby("Campaign Name").agg({"Purchases":"mean","Number of Impressions":"mean","Website Clicks":"mean","Content Viewed":"mean","Added to Cart":"mean"})

Purchases  Number of Impressions  Website Clicks  \
Campaign Name                                                         
Control Campaign  522.793103          109559.758621     5320.793103   
Test Campaign     521.233333           74584.800000     6032.333333   

                  Content Viewed  Added to Cart  
Campaign Name                                    
Control Campaign     1943.793103    1300.000000  
Test Campaign        1858.000000     881.533333

In [56]:
# import matplotlib.pyplot as plt
import plotly.graph_objects as go
from plotly.subplots import make_subplots

specs = [[{'type':'domain'}, {'type':'domain'},{'type':'domain'}], [{'type':'domain'}, {'type':'domain'},{'type':'domain'}]]
fig = make_subplots(rows=2, cols=3, specs=specs, subplot_titles=['Number of Impressions', 'Website Clicks',"Content Viewed",
                                                                 "Added to Cart","Amount Spent","Reach"])

# marker_colors = ["#2B3467", "#FFA3FD"]
marker_colors = ["gold","lightgreen"]

labels=["Control Campaign","Test Campaign"]
values= [sum(df_control["Number of Impressions"]),sum(df_test["Number of Impressions"])]
values1= [sum(df_control["Website Clicks"]),sum(df_test["Website Clicks"])]
values2= [sum(df_control["Content Viewed"]),sum(df_test["Content Viewed"])]
values3= [sum(df_control["Added to Cart"]),sum(df_test["Added to Cart"])]
values4= [sum(df_control["Amount Spent"]),sum(df_test["Amount Spent"])]
values5= [sum(df_control["Reach"]),sum(df_test["Reach"])]


fig.add_trace(go.Pie(labels=labels, values=values,
                     marker_colors=marker_colors),1,1)

fig.add_trace(go.Pie(labels=labels, values=values1,
                     marker_colors=marker_colors),1, 2)

fig.add_trace(go.Pie(labels=labels, values=values2,
                     marker_colors=marker_colors),1, 3)

fig.add_trace(go.Pie(labels=labels, values=values3,
                     marker_colors=marker_colors),2, 1)

fig.add_trace(go.Pie(labels=labels, values=values4,
                     marker_colors=marker_colors),2, 2)

fig.add_trace(go.Pie(labels=labels, values=values5,
                     marker_colors=marker_colors),2, 3)

fig.show()

 - The test campaign resulted in more searches on the website
 - The test campaign wins in the number of website clicks.
 - The audience of the control campaign viewed more content than the test campaign. Although there is not much difference, as the website clicks of the control campaign were low, its engagement on the website is higher than the test campaign.
 - Despite low website clicks more products were added to the cart from the control campaign.
 - The amount spent on the test campaign is higher than the control campaign. But as we can see that the control campaign resulted in more content views and more products in the cart, the control campaign is more efficient than the test campaign.
 - There’s only a difference of around 1% in the purchases made from both ad campaigns. As the Control campaign resulted in more sales in less amount spent on marketing, the control campaign wins here!

In [57]:
import plotly.graph_objects as go
import plotly.express as px
import plotly.io as pio

figure = px.scatter(data_frame = df, 
                    x="Amount Spent",
                    y="Number of Impressions", 
                    size="Amount Spent", 
                    color= "Campaign Name", 
                    trendline="ols")
figure.show()

Control group seems to have generated more impression based on the amount of money spent on campaigns. 

In [58]:
figure = px.scatter(data_frame = df, 
                    x="Added to Cart",
                    y="Purchases", 
                    size="Added to Cart", 
                    color= "Campaign Name", 
                    trendline="ols")
figure.show()

Control group seems to have higher numbers for added to cart, but the amount spent on purchases seems to be stagnant. Test campaign seems to have higher slope which indicates that the amount spent on purchases grows with the amount added to the cart. 

In [59]:
figure = px.scatter(data_frame = df, 
                    x="Content Viewed",
                    y="Added to Cart", 
                    size="Added to Cart", 
                    color= "Campaign Name", 
                    trendline="ols")
figure.show()


Control campaign seems to have more items added to the cart. 

# Hypothesis:

Null Hypothesis: There is no significant difference between the results of the campaigns

Alternative Hypothesis: There is a significant difference between the results of the campaigns (we want to specifically know if test groups performs better than control group).

Let's set significance level alpha as 5%.
Let's also asume that the objective of this analysis is to increase purchase. Therefore, we will conduct statistical testing on the Purchases column. We'll use t-test since the sample size is only 30.



###  Shapiro-Wilk test: checks whether the data comes from a normal distribution

 - Null Hypothesis: The data follows a normal distribution.
 - Alternative Hypothesis: The data does not follow a normal distribution.
The test computes two key values:

 - - Test Statistic (test_stat): Measures the degree to which the data deviates from a normal distribution. Values close to 1 indicate data that is close to normal.
 - - p-value: Indicates the probability of observing the data under the null hypothesis. A higher p-value suggests the null hypothesis (normality) is likely true.

In [60]:
test_stat,pvalue=shapiro(df.loc[df["Campaign Name"]=="Control Campaign","Purchases"])
print("Test Stat =%.4f, p-value=%.4f" % (test_stat,pvalue))

if pvalue < 0.05:  
    print("The 'Purchases' data for the 'Control Campaign' does not follow a normal distribution.")
else:
    print("The 'Purchases' data for the 'Control Campaign' follows a normal distribution.")


Test Stat =0.9433, p-value=0.1114
The 'Purchases' data for the 'Control Campaign' follows a normal distribution.


In [61]:
test_stat,pvalue=shapiro(df.loc[df["Campaign Name"]=="Test Campaign","Purchases"])
print("Test Stat =%.4f, p-value=%.4f" % (test_stat,pvalue))

if pvalue < 0.05:  
    print("The 'Purchases' data for the 'Test Campaign' does not follow a normal distribution.")
else:
    print("The 'Purchases' data for the 'Test Campaign' follows a normal distribution.")


Test Stat =0.9182, p-value=0.0241
The 'Purchases' data for the 'Test Campaign' does not follow a normal distribution.


### Finding P-value

In [62]:
test_stat,pvalue=ttest_ind(df.loc[df["Campaign Name"]=="Control Campaign","Purchases"],
                        df.loc[df["Campaign Name"]=="Test Campaign","Purchases"],equal_var=True)

print("Test Stat =%.4f, p-value=%.4f" % (test_stat,pvalue))

if pvalue < 0.05:
    print("There is a significant difference between the Test and Control Campaigns.")
    print("Null Hypothesis is rejected.")
else:
    print("There is no significant difference between the Test and Control Campaigns.")
    print("Null Hypothesis is not rejected.")

Test Stat =0.0307, p-value=0.9756
There is no significant difference between the Test and Control Campaigns.
Null Hypothesis is not rejected.


Since the data for the "Test Campaign" does not follow a normal distribution, using parametric tests (like a t-test) that assume normality might not be appropriate. So let's try something else:

**Using Mann-Whitney U Test**: A non-parametric alternative to the independent samples t-test, used to compare the medians of two independent groups.

In [63]:
control_data = df.loc[df["Campaign Name"] == "Control Campaign", "Purchases"]
test_data = df.loc[df["Campaign Name"] == "Test Campaign", "Purchases"]

stat, p_value = mannwhitneyu(control_data, test_data)
print("Stat =%.4f, p-value=%.4f" % (stat,p_value))

if p_value < 0.05:
    print("There is a significant difference between the Control and Test Campaigns.")
else:
    print("There is no significant difference between the Control and Test Campaigns.")


Stat =455.0000, p-value=0.9470
There is no significant difference between the Control and Test Campaigns.


**Interpretation**

Based on the results of the independent t-test and Mann-Whitney U Test, there is no statistically significant difference in the "Purchases" between the "Control Campaign" and the "Test Campaign" (p > 0.05). This suggests that the performance of the "Test Campaign" in terms of generating purchases is comparable to that of the "Control Campaign." Consequently, we do not have enough evidence to reject the null hypothesis, indicating that any observed difference in purchases between the two campaigns is likely due to random variation rather than a true effect of the campaign changes. Further analysis or additional data may be needed to determine if any other factors could influence the outcomes.

Let's look at other metrics: 
- **Click-Through Rate (CTR):** CTR measures the percentage of users who clicked on an ad or link out of the total number of impressions. It can indicate the effectiveness of the campaign in capturing users' attention and driving them to take initial action.

- **Conversion Rate:** Conversion Rate tracks the percentage of users who complete a desired action (like making a purchase) after clicking on an ad. It can indicate campaign's ability to turn clicks into actual outcomes, such as sales.

- **Cost Per Click (CPC):** CPC measures the cost incurred for each click on an ad. It describes efficiency of the campaign's spending in generating clicks and determining how cost-effective it is in driving traffic.

- **Return On Investment (ROI):** ROI calculates the profitability of the campaign by comparing the revenue generated to the cost of the campaign. It relates to the problem by determining the overall financial success and value derived from the campaign efforts.

In [64]:
df['CTR'] = (df['Website Clicks'] / df['Number of Impressions']) * 100
df['Conversion Rate'] = (df['Purchases'] / df['Website Clicks']) * 100
df['CPC'] = df['Amount Spent'] / df['Website Clicks']
df['ROI'] = ((df['Purchases'] - df['Amount Spent']) / df['Amount Spent']) * 100

df[['CTR', 'Conversion Rate','CPC','ROI']].head()

CTR  Conversion Rate       CPC        ROI
0  8.483471         8.808438  0.324971 -72.894737
1  6.700264         6.300863  0.216646 -70.916335
2  4.941121         5.716042  0.360018 -84.122919
3  4.205659        11.092985  0.632953 -82.474227
4  4.856521         9.825473  0.344873 -71.509913

In [65]:
df_KPIs=df.groupby("Campaign Name").agg({"CTR":"mean","Conversion Rate":"mean","CPC":"mean","ROI":"mean"})
df_KPIs

CTR  Conversion Rate       CPC        ROI
Campaign Name                                                    
Control Campaign   5.087893        11.422146  0.489907 -76.619613
Test Campaign     10.242260         9.231182  0.468718 -79.342253

In [66]:
#calculating each metric for each experiment group

df_test['CTR'] = df_test['Website Clicks'] / df_test['Number of Impressions']
df_control['CTR'] = df_control['Website Clicks'] / df_control['Number of Impressions']

df_control['Conversion Rate'] = df_control['Purchases'] /df_control['Website Clicks']
df_test['Conversion Rate'] = df_test['Purchases'] / df_test['Website Clicks']

df_control['CPC'] = df_control['Amount Spent'] /df_control['Website Clicks']
df_test['CPC'] = df_test['Amount Spent'] / df_test['Website Clicks']

df_control['ROI'] = ((df_control['Purchases'] - df_control['Amount Spent']) / df_control['Amount Spent']) * 100
df_test['ROI'] = ((df_test['Purchases'] - df_test['Amount Spent']) / df_test['Amount Spent']) * 100



In [67]:
metrics = ['CTR', 'Conversion Rate','CPC','ROI']

for i in metrics:
    t_stat, p_value = ttest_ind(df_control[i],df_test[i] )
    if p_value < 0.05:  
        print(i + ": there is significant difference.")
        print("p value : {}\n".format(p_value))
    else:
        print(i + ": no significant difference.")
        print("p value : {}\n".format(p_value))

CTR: there is significant difference.
p value : 0.00018398526312426124

Conversion Rate: no significant difference.
p value : 0.14214717143258024

CPC: no significant difference.
p value : 0.6827063408492927

ROI: no significant difference.
p value : 0.24146457998427984



The analysis of the campaign metrics reveals a mixed outcome. The Click-Through Rate (CTR) shows a statistically significant difference between the campaigns (p = 0.0002), indicating that test campaign is more effective in attracting user clicks. However, no significant differences were found in the Conversion Rate (p = 0.1421), Cost Per Click (CPC) (p = 0.6827), or Return on Investment (ROI) (p = 0.2415). This suggests that while test campaign may be better at capturing initial interest, it does not translate into more conversions, cost efficiency, or higher profitability compared to the other campaign. Further investigation may be needed to understand the disconnect between the high CTR and the lack of improvement in other key performance indicators.